### ISSUE TITLE
CERTIFICATE_VERIFY_FAILED on Own CA

### ISSUE THREAD

**Issue by skorebrits** [aws/aws-iot-device-sdk-python#33]
For my solution I want to use X.509 device certificates, as illustrated in http://docs.aws.amazon.com/iot/latest/developerguide/device-certs-your-own.html

When I'm trying to connect I get the following error:
SSLError(1, u'[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:661)')

I'm running:
certifi (2017.1.23)
cryptography (1.7.2)
AWSIoTPythonSDK (1.1.1)
boto3 (1.4.4)
botocore (1.5.11)
 
On python 2.7.13 with OpenSSL 1.0.2j.  I got X.509 working with via configure a device on the IOT section. But I want to generate the certificates. 

The Root CA is Generated via:
```bash
openssl genrsa -out rootCA.key 2048
root_subject="/C=NL/ST=City/L=City/O=Organisation/OU=Part/CN=Part"
openssl req -x509 -new -nodes -key rootCA.key -sha256 -days 1024 -out rootCA.pem -subj "$root_subject"

#Registration Code
aws_registration_code=$(aws iot get-registration-code)
registration_code=$(python -c "from sys import argv; import json;\
script, data = argv;\
data_dict = json.loads(data);\
print data_dict['registrationCode']" "$aws_registration_code")
echo "registration code: $registration_code"

#Key Pair
openssl genrsa -out verificationCert.key 2048
#CSR
csr_subject="/C=NL/ST=City/L=City/O=Organisation/OU=Part/CN=$registration_code"
echo $csr_subject
openssl req -new -key verificationCert.key -out verificationCert.csr -subj "$csr_subject"
#private key verification cert 
openssl x509 -req -in verificationCert.csr -CA rootCA.pem -CAkey rootCA.key -CAcreateserial -out verificationCert.pem -days 500 -sha256

#register with AWS
reg_response=$(aws iot register-ca-certificate --ca-certificate file://rootCA.pem --verification-cert file://verificationCert.pem)
cert_id=$(python -c "from sys import argv; import json; script, data = argv; item = json.loads(data); print item['certificateId']" "$reg_response")
aws iot update-ca-certificate --certificate-id $cert_id --new-status ACTIVE
```

The Device CA: 
```python
def create_certificate_on_disk(self):
        ''' Check if Certificate folder & certificates are available otherwise generate '''
        #check folder
        try:
            os.makedirs(self.device_cert_folder)
        except OSError:
            pass

        if not os.path.isfile(self.device_pem):
            #device key
            call(["openssl", "genrsa", "-out", self.device_key, "2048"])
            #device csr
            device_subject = "/C=NL/ST=City/L=City/O=Organisation/OU=Part/CN=%s" % self.identifier
            call(["openssl", "req", "-new", "-key", self.device_key, "-out", self.device_csr, "-subj", device_subject])
            #device pem
            call(["openssl", "x509", "-req", "-in", self.device_csr, "-CA", self.root_pem, "-CAkey", "root_ca/rootCA.key", "-CAcreateserial", "-out", self.device_pem, "-days", "500", "-sha256"])
        else:
            raise Exception("pem already created")
```

The device cert is then registered and set Active. 

The debug I get is:
```
2017-02-15 16:54:21,461 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Paho MQTT Client init.
2017-02-15 16:54:21,461 - AWSIoTPythonSDK.core.protocol.mqttCore - INFO - ClientID: basicPubSub
2017-02-15 16:54:21,461 - AWSIoTPythonSDK.core.protocol.mqttCore - INFO - Protocol: MQTTv3.1.1
2017-02-15 16:54:21,461 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Register Paho MQTT Client callbacks.
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - mqttCore init.
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Load CAFile from: root_ca/rootCA.pem
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Load Key from: device_certificates/deviceCert.key
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Load Cert from: device_certificates/deviceCert.pem
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Custom setting for backoff timing: baseReconnectTime = 1 sec
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Custom setting for backoff timing: maximumReconnectTime = 32 sec
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Custom setting for backoff timing: minimumConnectTime = 20 sec
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Custom setting for publish queueing: queueSize = -1
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Custom setting for publish queueing: dropBehavior = Drop Newest
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Custom setting for draining interval: 0.5 sec
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Set maximum connect/disconnect timeout to be 10 second.
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - DEBUG - Set maximum MQTT operation timeout to be 5 second
2017-02-15 16:54:21,462 - AWSIoTPythonSDK.core.protocol.mqttCore - INFO - Connection type: TLSv1.2 Mutual Authentication
2017-02-15 16:54:21,544 - AWSIoTPythonSDK.core - INFO - iOT Server error: SSLError(1, u'[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:661)')
```

I'm a missing a step?

**Comment by liuszeng** [comment]
Hi @skorebrits ,

Thank you very much for your interest in AWS IoT Python SDK.

We were able to duplicate the issue on our side by testing the credentials using Python device SDK and MQTT.fx. It seems that the server verification failed, which could be related to some mis-configuration for CA generation and registration. We already reached out the backend service team to check for further instructions and investigations. Sorry that we cannot give you a more positive answer for now. Will keep you updated.

Thanks,
Liusu

**Comment by skorebrits** [comment]
@liuszeng Thank you for the feedback. Hope they come with a swift solution.

**Comment by wenhuizhao** [comment]
I have the same problem. Any update on this?

**Comment by fengsongAWS** [comment]
Are you still using AWS IoT root CA? Just note that from the instructions, "You should use the AWS IoT root CA certificate when connecting to AWS IoT even if you register your own root CA certificate. The AWS IoT root CA certificate is used by a device to verify the identity of the AWS IoT servers. "

Thanks

**Comment by wenhuizhao** [comment]
I was using my own root CA before. After I change to use AWS IoT root CA certificate. I got a different error: 
Googled the error. Looks like invalid CA certificate. 


    self.client.connect()
  File "build/bdist.macosx-10.12-intel/egg/AWSIoTPythonSDK/MQTTLib.py", line 403, in connect
  File "build/bdist.macosx-10.12-intel/egg/AWSIoTPythonSDK/core/protocol/mqttCore.py", line 290, in connect
  File "build/bdist.macosx-10.12-intel/egg/AWSIoTPythonSDK/core/protocol/paho/client.py", line 655, in connect
  File "build/bdist.macosx-10.12-intel/egg/AWSIoTPythonSDK/core/protocol/paho/client.py", line 798, in reconnect
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ssl.py", line 943, in wrap_socket
    ciphers=ciphers)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ssl.py", line 611, in __init__
    self.do_handshake()
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ssl.py", line 840, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: SSLV3_ALERT_CERTIFICATE_UNKNOWN] sslv3 alert certificate unknown (_ssl.c:661)

**Comment by fengsongAWS** [comment]
Hi @wenhuizhao ,

Did you register your CA and sign your device certificate? Could you elaborate your steps how you register your certificates? 
You can try following instructions to see if you have any problems. 
http://docs.aws.amazon.com/iot/latest/developerguide/device-certs-your-own.html

Thanks

**Comment by wenhuizhao** [comment]
Yes. I did register my CA and sign my device certificate. Step is
1. openssl genrsa -out rootCA.key 2048   
2. openssl req -x509 -new -nodes -key rootCA.key -sha256 -days 1024 -out rootCA.pem
3. aws iot get-registration-code
4. openssl genrsa -out verificationCert.key 2048
5. openssl req -new -key verificationCert.key -out verificationCert.csr
(using my registration code as Common Name)
6. openssl x509 -req -in verificationCert.csr -CA rootCA.pem -CAkey rootCA.key -CAcreateserial -out verificationCert.pem -days 500 -sha256
7. aws iot register-ca-certificate --ca-certificate file://rootCA.pem --verification-cert file://verificationCert.pem  
8. aws iot update-ca-certificate --certificate-id xxxxxxxxxxx --new-status ACTIVE
9. aws iot update-ca-certificate --certificate-id xxxxxxxxxxx --new-auto-registration-status ENABLE
I can see the CA certificate in my IOT console as active and Auto-Registration enabled. 
Generate device certificate:
1. openssl genrsa -out deviceCert.key 2048
2. openssl req -new -key deviceCert.key -out deviceCert.csr
3. openssl x509 -req -in deviceCert.csr -CA rootCA.pem -CAkey rootCA.key -CAcreateserial -out deviceCert.pem -days 500 -sha256

Next I use the deviceCert.key and deviceCert.pem ,  and amazon  iot root CA in my python code. It 
give me error in my previous post.

**Comment by wenhuizhao** [comment]
BTW, I am using us-west-2 for all my certificate and sdk code configure.

**Comment by vatjjar** [comment]
I stumbled today to this exact same problem with the same symptoms. I am using too us-west-2. Any further info to this from AWS team?

**Comment by wenhuizhao** [comment]
I haven't heard anything back from AWS team.

**Comment by vatjjar** [comment]
Maybe it will help this thread, if I describe my steps too.

First, based on information in https://aws.amazon.com/blogs/mobile/use-your-own-certificate-with-aws-iot/ I generate my keys:

1. openssl genrsa -out sampleCACertificateOne.key 2048
2. openssl req -x509 -new -nodes -key sampleCACertificateOne.key -sha256 -days 365 -out sampleCACertificateOne.pem
3.  ... at this point grab the registration code ...
4. openssl genrsa -out privateKeyVerificationOne.key 2048
5. openssl req -new -key privateKeyVerificationOne.key -out privateKeyVerificationOne.csr
6. openssl x509 -req -in privateKeyVerificationOne.csr -CA sampleCACertificateOne.pem -CAkey sampleCACertificateOne.key -CAcreateserial -out privateKeyVerificationOne.crt -days 365 -sha256
7. ... I insert the registration code to "Common Name" when asked by openssl

Then I go to IoT console, and upload my CS using files sampleCACertificateOne.pem and privateKeyVerificationOne.crt. No problem at this point.

Then I programmatically register a new thing using steps:

1. iot.create_thing(<params>)
2. iot.describe_endpoint()
3. iot.create_keys_and_certificates(<params>)
4. iot.attach_principal_policy(<params>)
5. iot.attach_thing_principal(<params>)

And of course, save all keys and credentials for later use. All these steps work 100% fine. I get a new thing, I see it in IoT control panel with its certificates, etc. Everything seems to be just fine.

Then I try to start MQTT client and connect to AWS IoT broker. AWSIoTMQTTClient.configureCredentials() require root CA (I use sampleCACertificateOne.pem), private key and thing certificate (both come in create_keys_and_certificates() call earlier). Connection request fails to:

2017-05-10 10:23:09,063 - AWSIoTPythonSDK.core.protocol.mqttCore - INFO - Connection type: TLSv1.2 Mutual Authentication
ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:645)

At this point, it is worth noting that if I download root CA from https://www.symantec.com/content/en/us/enterprise/verisign/roots/VeriSign-Class%203-Public-Primary-Certification-Authority-G5.pem and run the whole process with exact same steps, the SYSTEM WORKS!

Hence, I suspect that this is related to the key generation process itself (maybe wrong params), or a bug in AWS IoT managing the keys.

I hope this helps in any possible way...

**Comment by vatjjar** [comment]
I managed to get my issue resolved. In my implementation I had two problems:

1) In the steps one sees that I have "3. iot.create_keys_and_certificate()". I'm not sure is this supposed to work when using custom root CA or not, but nevertheless I now create device keys and certificates locally and sign them with my root CA. Then I replace that particular step with: iot.register_certificate(certificatePem="mydevicecertpem", caCertificatePem="myrootcapem", setAsActive=True). In other words, I pass the whole device certs/keys as one lump to AWS instead asking aws to generate them for me.

2) Finally, I think the actual CERTIFICATE_VERIFY_FAILED came from the Paho client trying to register to AWS MQTT broker. One needs to pass rootCA.pem for it, BUT IT NEEDS TO BE THE AWS ROOT CA PEM.  I was mislead in this some why, and did not get that during the mqttclient.connect() the given root CA is used to verify the AWS broker authenticity, NOT the thing certificate. I was blinded in getting rid of the amazon root CA altogether and hence I tried to use my root CA in this step, which obviously fails.

Given these two fixes, my thing communication now works with my on custom root CA and custom device certificates/keys.

**Comment by liuszeng** [comment]
Hi everyone,

Thanks for sharing all the information here. The following blogs describe the details about using self-signed CA/Device certificate and Just-In-Time-Registration (JITR) for device certificates:
https://aws.amazon.com/blogs/mobile/use-your-own-certificate-with-aws-iot/
https://aws.amazon.com/blogs/iot/just-in-time-registration-of-device-certificates-on-aws-iot/

Some take-away:
- Custom CA needs to be registered to AWS IoT with the correct registration code configured and be activated.
- Device certificate signed by the CA needs to be uploaded to AWS IoT with the correct CName configured for connecting to AWS IoT your custom AWS IoT MQTT endpoint.
- As always, device certificate needs to have a proper IoT policy attached to it to allow proper authorization, which can be achieved automatically by using JITR.
- When connecting to AWS IoT using custom device certificate, you will still need to use AWS IoT root CA, which is used by the device to verify the AWS IoT server identity it is trying to connect to.

Sorry for the inconvenience that has been caused here. Hope the above helps.

Thanks,
Liusu

**Comment by NguyenMinhKhanhBK** [comment]
I had the same problem. I followed this instruction to generate my own certificate (rootCA.pem, deviceCert.pem, deviceCert.key). 
[(https://docs.aws.amazon.com/iot/latest/developerguide/device-certs-your-own.html)]
Everything works fine, but when I use these files to connect to AWS (via MQTT.fx client), I get "**Invalid key format**" notification. 
I don't know what happen and how to fix it. Please help me!
BTW, the PLC programming software I'm using (TIA Portal) doesn't allow to import **.key** type. So I can not connect to AWS using the certificates generated by AWS IoT (as it includes client.key file). I combined the **client.cert** and **client.key** to a single file named **client.pem** but nothing works. I hope someone will help me to solve these problems. Thank you

**Comment by tutika** [comment]
I am also facing the similar issue when downloaded the certificates and using it in MQTT.fx client.
The same certificates are working with other IoT device.

**Comment by gopi-ideas2it** [comment]
Hi all,

**Comment by gopi-ideas2it** [comment]
I have fixed the issue by updating openssl to 1.0.1t
